tls_manager.go

```diff
@@ -0,0 +1,272 @@
+		}
+	}
+
+	var reloadedCertData tls.Certificate
+	// If the certificate expired or it was outdated, delete it and the TLS
+	// key and generate a new pair.
+	if time.Now().After(parsedCert.NotAfter) || refresh {
+		ltndLog.Info("TLS certificate is expired or outdated, " +
+			"generating a new one")
+
+		err := os.Remove(t.cfg.TLSCertPath)
+		if err != nil {
+			return tls.Certificate{}, err
+		}
+
+		err = os.Remove(t.cfg.TLSKeyPath)
+		if err != nil {
+			return tls.Certificate{}, err
+		}
+
+		rpcsLog.Infof("Renewing TLS certificates...")
+		err = cert.GenCertPair(
+			"lnd autogenerated cert", t.cfg.TLSCertPath,
+			t.cfg.TLSKeyPath, t.cfg.TLSExtraIPs,
+			t.cfg.TLSExtraDomains, t.cfg.TLSDisableAutofill,
+			t.cfg.TLSCertDuration,
+		)
+		if err != nil {
+			return tls.Certificate{}, err
+		}
+		rpcsLog.Infof("Done renewing TLS certificates")
+
+		// Reload the certificate data.
+		reloadedCertData, _, err = cert.LoadCert(
+			t.cfg.TLSCertPath, t.cfg.TLSKeyPath,
+		)
+		if err != nil {
+			return tls.Certificate{}, err
+		}
+	}
+
+	return reloadedCertData, nil
+}
+
+// setUpLetsEncrypt automatically generates a Let's Encrypt certificate if the
+// option is set.
+func (t *TLSManager) setUpLetsEncrypt(certData *tls.Certificate,
+	tlsCfg *tls.Config) func() {
+
+	// If Let's Encrypt is enabled, instantiate autocert to request/renew
+	// the certificates.
+	cleanUp := func() {}
+	if t.cfg.LetsEncryptDomain != "" {
+		ltndLog.Infof("Using Let's Encrypt certificate for domain %v",
+			t.cfg.LetsEncryptDomain)
+
+		manager := autocert.Manager{
+			Cache:  autocert.DirCache(t.cfg.LetsEncryptDir),
+			Prompt: autocert.AcceptTOS,
+			HostPolicy: autocert.HostWhitelist(
+				t.cfg.LetsEncryptDomain,
+			),
+		}
+
+		srv := &http.Server{
+			Addr:    t.cfg.LetsEncryptListen,
+			Handler: manager.HTTPHandler(nil),
+		}
+		shutdownCompleted := make(chan struct{})
+		cleanUp = func() {
+			err := srv.Shutdown(context.Background())
+			if err != nil {
+				ltndLog.Errorf("Autocert listener shutdown "+
+					" error: %v", err)
+
+				return
+			}
+			<-shutdownCompleted
+			ltndLog.Infof("Autocert challenge listener stopped")
+		}
+
+		go func() {
+			ltndLog.Infof("Autocert challenge listener started "+
+				"at %v", t.cfg.LetsEncryptListen)
+
+			err := srv.ListenAndServe()
+			if err != http.ErrServerClosed {
+				ltndLog.Errorf("autocert http: %v", err)
+			}
+			close(shutdownCompleted)
+		}()
+
+		getCertificate := func(h *tls.ClientHelloInfo) (
+			*tls.Certificate, error) {
+
+			lecert, err := manager.GetCertificate(h)
+			if err != nil {
+				ltndLog.Errorf("GetCertificate: %v", err)
+				return certData, nil
+			}
+
+			return lecert, err
+		}
+
+		// The self-signed tls.cert remains available as fallback.
+		tlsCfg.GetCertificate = getCertificate
+	}
+
+	return cleanUp
+}
+
+// fileExists reports whether the named file or directory exists.
+// This function is taken from https://github.com/btcsuite/btcd
+func fileExists(name string) bool {
```

Comment:
Can just use `lnrpc.FileExists` instead, since we have the same function in 4 different places otherwise.

Reply:
Sounds good. Since this function was used in other files in the main lnd package, I moved this change into a separate, small commit for more clarity.